### PR TITLE
feat: support theme default config and extra pages rendering

### PR DESCRIPTION
**Pull request by meteorlxy**
Three features:
1. Support theme default config by `theme-dir/config.js`
2. Allow custom theme to render extra pages ([#198](https://github.com/vuejs/vuepress/issues/198#issuecomment-383378566)) (Another implementation #313)
3. Allow user to render extra pages based on a theme (#160)

Feature 1 is necessary for Feature 2. I propose to use `theme-dir/config.js`, which is similar to `.vuepress/config.js`, as the theme default config

To achieve Feature 2,  add `extraPages` in `theme-dir/config.js`, and add coresponding extra routes in `theme-dir/enhanceApp.js`. That will enable extra pages rendering for theme.

To achieve Feature 3, add `extraPages` in `.vuepress/config.js`, and add coresponding extra routes in `.vuepress/enhanceApp.js`. That will enable extra pages rendering based on a theme. Notice that theme's `extraPages` config could be overwritten in `.vuepress/config.js`.

E.g: 
```js
// .vuepress/config.js
// or theme-dir/config.js
module.exports = {
  extraPages: [
    '/test'
  ],
  // other config...
}
```

```js
// .vuepress/enhanceApp.js
// or theme-dir/enhanceApp.js
const Test = () => import("./Test");

export default ({ Vue, options, router }) => {
  router.addRoutes([
    { path: "/test", component: Test }
  ])
}
```

Docs will be added after accepted.

### CODE REVIEW

**Comment by meteorlxy** [comment]
What do you think about this? @mdaffin

File: lib/prepare.js [hunk 1]
```diff
@@ -335,6 +324,26 @@ function sort (arr) {
...
 
+async function resolveConfig (dir) {
+  const configPath = path.resolve(dir, 'config.js')
+  const configYmlPath = path.resolve(dir, 'config.yml')
+  const configTomlPath = path.resolve(dir, 'config.toml')
+
+  delete require.cache[configPath]
+
+  // resolve siteConfig
+  let config = {}
```
Comment (ycmjason) [inline lib/prepare.js]: Just my honest opinion 😊❤️🙈, I think 335-344 would be cleaner with the following: 

```js
if (fs.existsSync(configYmlPath)) {
  return await parseConfig(configYmlPath)
}

if (fs.existsSync(configTomlPath)) {
  return await parseConfig(configTomlPath)
}

if (fs.existsSync(configPath)) {
  return require(configPath)
}

return {}
```

Comment (meteorlxy) [reply lib/prepare.js]: Uh looks better. I just moved things down into a functions without much refactor.

Comment (ycmjason) [reply lib/prepare.js]: Yea yea! Just spot that when I see the changes, might be worth it to make the change.

**Comment by meteorlxy** [comment]
@ycmjason Maybe we should ensure all the files have been rendered? Or some of the pages may not been rendered before the script exits?

**Comment by meteorlxy** [comment]
@ycmjason Yeah, I also thought about `Promise.all` but didn't try it :smile:

**Comment by ycmjason** [comment]
@meteorlxy 
https://github.com/vuejs/vuepress/commit/9719bc3246afcf77f69eaf94a34f0d6b2fa9c95b

Looks like the for-await loop is intentional. I wonder why tho.

**Comment by ycmjason** [comment]
I see, this is probably for better CLI output. The time is not much different so Evan chose to use for-await probably..

**Comment by Feofilakt** [comment]
Why is it closed?

**Pull request closed without merging**